File: GO_Backend/internal/queue/publisher.go
package queue

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/rs/zerolog/log"
	amqp "github.com/rabbitmq/amqp091-go"
	"github.com/your-org/alumni-go/internal/crypto"
)

// Publish sends a message to the alumni.exchange with the given routing key.
// The message body is JSON-encoded and HMAC-signed if a secret is provided.
func Publish(c *Channel, routingKey string, payload interface{}, hmacSecret string) error {
	body, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("failed to marshal message: %w", err)
	}

	headers := amqp.Table{
		"timestamp": time.Now().Unix(),
	}

	// Sign the message with HMAC if secret is configured
	if hmacSecret != "" {
		signature := crypto.SignMessage(body, hmacSecret)
		headers["x-signature"] = signature
	}

	ch := c.GetChannel()
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	err = ch.PublishWithContext(ctx,
		"alumni.exchange", // exchange
		routingKey,        // routing key
		false,             // mandatory
		false,             // immediate
		amqp.Publishing{
			ContentType:  "application/json",
			DeliveryMode: amqp.Persistent,
			Timestamp:    time.Now(),
			Headers:      headers,
			Body:         body,
		})
	if err != nil {
		return fmt.Errorf("failed to publish message: %w", err)
	}

	log.Debug().Str("routingKey", routingKey).Int("bodyLen", len(body)).Msg("Message published")
	return nil
}

// PublishRaw sends raw bytes to the exchange with optional HMAC signing.
func PublishRaw(c *Channel, routingKey string, body []byte, hmacSecret string) error {
	headers := amqp.Table{
		"timestamp": time.Now().Unix(),
	}

	if hmacSecret != "" {
		headers["x-signature"] = crypto.SignMessage(body, hmacSecret)
	}

	ch := c.GetChannel()
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	return ch.PublishWithContext(ctx,
		"alumni.exchange",
		routingKey,
		false,
		false,
		amqp.Publishing{
			ContentType:  "application/json",
			DeliveryMode: amqp.Persistent,
			Timestamp:    time.Now(),
			Headers:      headers,
			Body:         body,
		})
}

File: pipeline/publisher.py
"""
Shared RabbitMQ publisher for the Python pipeline.
All pipeline scripts use this module to publish events with HMAC signing.
"""

import os
import json
import hmac
import hashlib
import time
import pika
import logging

logger = logging.getLogger(__name__)

RABBITMQ_URL = os.environ.get('RABBITMQ_URL', 'amqp://user:password@localhost:5672/')
HMAC_SECRET = os.environ.get('INTERNAL_HMAC_SECRET', '')


def _get_connection():
    """Create a new blocking connection to RabbitMQ."""
    parameters = pika.URLParameters(RABBITMQ_URL)
    parameters.heartbeat = 600
    parameters.blocked_connection_timeout = 300
    return pika.BlockingConnection(parameters)


def sign_message(body: bytes) -> str:
    """Create HMAC-SHA256 signature for a message body."""
    if not HMAC_SECRET:
        return ''
    return hmac.new(
        HMAC_SECRET.encode(), body, hashlib.sha256
    ).hexdigest()


def publish_event(routing_key: str, payload: dict):
    """
    Publish a JSON message to alumni.exchange with HMAC signing.

    Args:
        routing_key: RabbitMQ routing key (e.g., 'import.enriched')
        payload: Dictionary to JSON-encode and publish.
    """
    body = json.dumps(payload).encode('utf-8')
    signature = sign_message(body)

    headers = {
        'x-signature': signature,
        'x-timestamp': int(time.time()),
        'x-source': 'python-pipeline',
    }

    try:
        connection = _get_connection()
        channel = connection.channel()
        channel.exchange_declare(
            exchange='alumni.exchange',
            exchange_type='topic',
            durable=True
        )
        channel.basic_publish(
            exchange='alumni.exchange',
            routing_key=routing_key,
            body=body,
            properties=pika.BasicProperties(
                delivery_mode=2,  # persistent
                content_type='application/json',
                headers=headers,
            )
        )
        connection.close()
        logger.info(f"Published to {routing_key}: {json.dumps(payload)[:100]}...")
    except Exception as e:
        logger.error(f"Failed to publish to {routing_key}: {e}")
        raise


def publish_batch(routing_key: str, payloads: list, batch_size: int = 50):
    """
    Publish multiple messages in a single connection.

    Args:
        routing_key: RabbitMQ routing key.
        payloads: List of dictionaries to publish.
        batch_size: Number of messages per connection.
    """
    try:
        connection = _get_connection()
        channel = connection.channel()
        channel.exchange_declare(
            exchange='alumni.exchange',
            exchange_type='topic',
            durable=True
        )

        for i, payload in enumerate(payloads):
            body = json.dumps(payload).encode('utf-8')
            signature = sign_message(body)

            channel.basic_publish(
                exchange='alumni.exchange',
                routing_key=routing_key,
                body=body,
                properties=pika.BasicProperties(
                    delivery_mode=2,
                    content_type='application/json',
                    headers={
                        'x-signature': signature,
                        'x-timestamp': int(time.time()),
                        'x-source': 'python-pipeline',
                    },
                )
            )

            # Reconnect every batch_size messages to avoid timeouts
            if (i + 1) % batch_size == 0:
                connection.close()
                connection = _get_connection()
                channel = connection.channel()

        connection.close()
        logger.info(f"Published {len(payloads)} messages to {routing_key}")
    except Exception as e:
        logger.error(f"Batch publish failed: {e}")
        raise
